spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,415 @@
+                if len(prospective_group) > 0:
+                    if k == 0:
+                        group = prospective_group
+                    break
+                # Only keep going if all n still match on the kth token
+                _, sample_toks = next(group.iteritems())
+                prospective_group = {idx: toks for idx, toks
+                                     in group.iteritems()
+                                     if toks[k] == sample_toks[k]}
+                if len(prospective_group) == len(group) or k == 0:
+                    group = prospective_group
+                    k += 1
+                else:
+                    break
+            _, sample_toks = next(group.iteritems())
+            if k == 0:
+                short_form = ''
+            if k == 1:
+                short_form = sample_toks[0]
+            elif k == 2:
+                short_form = sample_toks[0] + os.sep + sample_toks[1]
+            elif k > 2:
+                short_form = sample_toks[0] + "..." + os.sep + sample_toks[k-1]
+            for idx in group.keys():
+                new_path_list[idx] += short_form + (os.sep if k > 0 else '')
+                del level_idx[idx]
+            recurse_level({idx: toks[k:] for idx, toks in group.iteritems()})
+
+    recurse_level(dict(enumerate(path_list)))
+    return [path[:-1] for path in new_path_list]  # trim final sep
+
+
+class HTMLDelegate(QStyledItemDelegate):
+    """With this delegate, a QListWidgetItem can render HTML.
+
+    Taken from http://stackoverflow.com/a/5443112/2399799
+    """
+    def paint(self, painter, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        style = QApplication.style() if options.widget is None\
+            else options.widget.style()
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+
+        options.text = ""
+        style.drawControl(QStyle.CE_ItemViewItem, options, painter)
+
+        ctx = QAbstractTextDocumentLayout.PaintContext()
+
+        textRect = style.subElementRect(QStyle.SE_ItemViewItemText, options)
+        painter.save()
+        painter.translate(textRect.topLeft())
+        painter.setClipRect(textRect.translated(-textRect.topLeft()))
+        doc.documentLayout().draw(painter, ctx)
+
+        painter.restore()
+
+    def sizeHint(self, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+        doc.setTextWidth(options.rect.width())
+        return QSize(doc.idealWidth(), doc.size().height())
+
+
+class UpDownFilter(QObject):
+    """Use with ``installEventFilter`` to get up/down arrow key press signal.
+    """
+    up_down = Signal(int)
+
+    def eventFilter(self, src, e):
+        if isinstance(e, QKeyEvent) and e.type() == QEvent.KeyPress:
```

Comment:
Why is the `isinstance()` check needed ? `e` will always be an event and have a `.type()``